src/Rheology.jl

```diff
@@ -0,0 +1,210 @@
+function StrainRateTrial_Invariant(τII, G, Δt, B, n)
+    ε̇II_vis   = B.*τII.^n 
+    ε̇II_trial = ε̇II_vis + τII/(2*G*Δt)
+    return ε̇II_trial
+end
+
+function LocalRheology(ε̇, materials, phases, Δ)
+
+    # Effective strain rate & pressure
+    ε̇II  = sqrt.(1/2*(ε̇[1]^2 + ε̇[2]^2 + (-ε̇[1]-ε̇[2])^2) + ε̇[3]^2)
+    P    = ε̇[4]
+
+    # Parameters
+    ϵ    = 1e-10 # tolerance
+    n    = materials.n[phases]
+    η0   = materials.η0[phases]
+    B    = materials.B[phases]
+    G    = materials.G[phases]
+    C    = materials.C[phases]
+    ϕ    = materials.ϕ[phases]
+    ηvp  = materials.ηvp[phases]
+    ψ    = materials.ψ[phases]    
+    β    = materials.β[phases]
+    comp = materials.compressible
+
+    # Initial guess
+    η    =  (η0 .* ε̇II.^(1 ./ n .- 1.0 ))[1]
+    ηvep = inv(1/η + 1/(G*Δ.t))
+    τII  = 2*ηvep*ε̇II
+
+    # Visco-elastic powerlaw
+    for it=1:20
+        r      = ε̇II - StrainRateTrial_Invariant(τII, G, Δ.t, B, n)
+        # @show abs(r)
+        (abs(r)<ϵ) && break
+        ∂ε̇II∂τII = Enzyme.jacobian(Enzyme.Forward, StrainRateTrial_Invariant, τII, G, Δ.t, B, n)
+        ∂τII∂ε̇II = inv(∂ε̇II∂τII[1])
+        τII     += ∂τII∂ε̇II*r
+    end
+    isnan(τII) && error()
+ 
+    # Viscoplastic return mapping
+    λ̇    = 0.0
+    F    = τII - C*cosd(ϕ) - P*sind(ϕ )- λ̇*ηvp
+    if F > 1e-10
+        λ̇    = F / (ηvep + ηvp + comp*Δ.t/β*sind(ϕ)*sind(ψ)) 
+        τII -= λ̇ * ηvep
+        P   += comp * λ̇  * sind(ψ) * Δ.t / β
+        ηvep = τII/(2*ε̇II)
+
+        # τII = C*cosd(ϕ) + P*sind(ϕ) + ηvp*λ̇
+        F    = τII - C*cosd(ϕ) - P*sind(ϕ )- λ̇*ηvp
+        (F>1e-10) && error("Failed return mapping")
+        (τII<0.0) && error("Plasticity without condom")
+    end
+
+    # Effective viscosity
+    ηvep = τII/(2*ε̇II)
+
+    return ηvep, λ̇, P
+end
+
+function StressVector!(ε̇, materials, phases, Δ) 
+    η, λ̇, P = LocalRheology(ε̇, materials, phases, Δ)
+    τ       = @SVector([2 * η * ε̇[1],
+                        2 * η * ε̇[2],
+                        2 * η * ε̇[3],
+                                  P])
+    return τ, η, λ̇
+end
+
+function TangentOperator!(𝐷, 𝐷_ctl, τ, τ0, ε̇, λ̇, η , V, Pt, ΔPt, type, BC, materials, phases, Δ)
+
+    _ones = @SVector ones(4)
+
+    # Loop over centroids
+    for j=1:size(ε̇.xx,2)-0, i=1:size(ε̇.xx,1)-0
+        if (i==1 && j==1) || (i==size(ε̇.xx,1) && j==1) || (i==1 && j==size(ε̇.xx,2)) || (i==size(ε̇.xx,1) && j==size(ε̇.xx,2))
+            # Avoid the outer corners - nothing is well defined there ;)
+        else
```

Comment:
```suggestion
        if !((i==1 && j==1) || (i==size(ε̇.xx,1) && j==1) || (i==1 && j==size(ε̇.xx,2)) || (i==size(ε̇.xx,1) && j==size(ε̇.xx,2))) # Avoid the outer corners - nothing is well defined there ;)
```
The empty branch can be eliminated